### PR TITLE
[c++] Implements ArraySort/ToSet/ToDict in Emit.

### PR DESCRIPTION
**Pull request by catoverdrive**
Builds on #5463.

This currently just pulls values out into a vector and sorts the vector based on a defined sorting function; could possibly be improved in the future by defining an iterator that we can call e.g. `std::stable_sort` on.

### CODE REVIEW

File: hail/src/main/resources/include/hail/ArraySorter.h [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+    template<typename IsEqual>
+    void distinct() {
+      IsEqual eq_;
+      for (auto it = ++non_missing_.begin(); it != non_missing_.end(); ) {
+        if (eq_((*it).elem_, (*(it - 1)).elem_)) {
+            it = non_missing_.erase(it);
+        } else { ++it; }
+      }
+      if (n_missing_ != 0) { n_missing_ = 1; }
```
Comment (patrick-schultz) [inline hail/src/main/resources/include/hail/ArraySorter.h]: Calling `erase` repeatedly in the middle of the vector has potential quadratic complexity (it has to shift down everything above each time). Better to use
```C++
auto new_end = std::unique(non_missing_.begin(), non_missing_.end(), eq_);
non_missing_.erase(new_end, non_missing_.end())
```

File: hail/src/main/resources/include/hail/ArraySorter.h [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+template <typename ArrayBuilder, typename LessThan>
+class ArraySorter {
+  public:
+    using ArrayImpl = typename ArrayBuilder::ArrayImpl;
+    using T = typename ArrayImpl::T;
+
+    struct SortableElem {
+      T elem_;
+      LessThan lt_;
```
Comment (patrick-schultz) [inline hail/src/main/resources/include/hail/ArraySorter.h]: I don't think we should be embedding a copy of the comparison functor into every element. If `LessThan` happened to have non-zero size, we're now inflating the entire collection. I think the following should work fine:
```c++
std::stable_sort(non_missing_.begin(), non_missing_.end(), LessThan())
```

**Comment by catoverdrive** [comment]
@patrick-schultz this is tested in the IRSuite tests for ArraySort/ToSet/ToDict. Do you want c++ tests for the ArraySorter class?

**Comment by patrick-schultz** [comment]
Ah great. I just wanted to make sure the new code was getting exercised by the IR tests.

**Pull request merged**